contracts/src/contracts/nullifier_registry/nullifier_registry.cairo

```diff
@@ -1 +1,294 @@
+#[starknet::contract]
+pub mod NullifierRegistry {
+    use starknet::{
+        ContractAddress, get_caller_address, get_block_timestamp,
+        storage::{Map, StorageMapReadAccess, StorageMapWriteAccess}
+    };
+    use zkramp::contracts::nullifier_registry::interface::{
+        INullifierRegistry, INullifierRegistryDispatcher, INullifierRegistryDispatcherTrait
+    };
+
+    #[storage]
+    pub struct Storage {
+        // / map(index -> writers address)
+        writers: Map<u64, ContractAddress>,
+        writers_count: u64,
+        is_nullified: Map<u256, bool>,
+        is_writers: Map<ContractAddress, bool>
+    }
+
+
+    #[event]
+    #[derive(Drop, starknet::Event)]
+    pub enum Event {
+        NullifierAdded: NullifierAdded,
+        WriterAdded: WriterAdded,
+        WriterRemoved: WriterRemoved
+    }
+
+    /// @notice Emitted when the account is locked
+    /// @param account tokenbound account who's lock function was triggered
+    /// @param locked_at timestamp at which the lock function was triggered
+    /// @param lock_until time duration for which the account remains locked in second
+    #[derive(Drop, starknet::Event)]
+    pub struct NullifierAdded {
+        #[key]
+        pub nullifier: u256,
+        pub writer: ContractAddress
+    }
+    #[derive(Drop, starknet::Event)]
+    pub struct WriterAdded {
+        #[key]
+        pub writer: ContractAddress
+    }
+    #[derive(Drop, starknet::Event)]
+    pub struct WriterRemoved {
+        #[key]
+        pub writer: ContractAddress
+    }
+
+
+    #[constructor]
+    fn constructor(ref self: ContractState) {
+        self.writers_count.write(0);
+    }
+
+    #[abi(embed_v0)]
+    impl NullifierRegistryImpl of INullifierRegistry<ContractState> {
+        fn add_nullifier(ref self: ContractState, nullifier: u256) {
+            assert(self.is_writer(get_caller_address()), 'Caller is not a writer');
+            assert(!self.is_nullified(nullifier), 'Nullifier already exists');
+
+            self.is_nullified.write(nullifier, true);
+            // emit event here
+            self.emit(NullifierAdded { nullifier: nullifier, writer: get_caller_address(), });
+        }
+
+
+        fn add_write_permissions(ref self: ContractState, new_writer: ContractAddress) {
```

Comment:
This is only allowed for the contract owner, you should use [`Ownable`](https://github.com/OpenZeppelin/cairo-contracts/blob/main/packages/access/src/ownable/ownable.cairo) similarly to how it's done in the original